### DISCUSSION TITLE
VSCode keeps switching the language of my files to C++

### DISCUSSION THREAD

**Discussion by Zylann** [microsoft/vscode-cpptools#12136] [Q&A]
I'm not sure where to ask this, but recently my VSCode started to switch the language of many different file types to C++. XML files are now C++, JSON files are C++, even TXT... and they get filled with squiggles and get wrong highlighting etc. It's like VSCode no longer recognizes them as they should be and somehow decides to fallback on C++.

![image](https://github.com/microsoft/vscode-cpptools/assets/1311555/139b9ae3-6f5b-424a-ac71-356fdca93dfa)
![image](https://github.com/microsoft/vscode-cpptools/assets/1311555/5bd44732-de37-4c72-9cf0-f1f68cbef157)
![image](https://github.com/microsoft/vscode-cpptools/assets/1311555/eba11555-7ea9-436d-bd6c-73062b5d180e)

I have to manually switch them back to what they should be, but they eventually switch back to C++, if I restart VSCode for example.
Any idea what's going on?

------

Ok, I eventually found this at the end of my `settings.json`:
```json
        "strstream": "cpp",
        "syncstream": "cpp",
        "typeindex": "cpp",
        "*.idl": "cpp",
        "*.odl": "cpp",
        "*.xsd": "cpp",
        "*.json": "cpp",
        "*.txt": "cpp",
        "sconstruct": "cpp",
        "scsub": "cpp",
        "*.py": "cpp",
        "*.zip": "cpp",
        "*.xml": "cpp",
        "*.pdf": "cpp",
        "*.webp": "cpp",
        "*.pdn": "cpp",
        "*.ods": "cpp",
        "*.blend": "cpp",
        "*.schematic": "cpp",
        "*.gd": "cpp",
        "*.kt": "cpp",
        "*.dat": "cpp",
        "*.sh": "cpp"
    },
```

I have no idea what put these here, but it wasn't me :| Maybe a buggy extension? The last change was from a few hours ago, I only have cpptools, cmake, git blame, meson, shader languages and zig

**Comment by bobbrow** [comment]
I think an extension is responsible for this.  Our extension does not add associations for file extensions.  We only add associations for STL headers that have no extension and don't get associated with cpp.  We do have a setting to turn off that behavior though.

**Reply by sean-mcmanus** [reply]
@bobbrow Actually, our extension does add the `"*.<ext>"` ("by design", e.g. when using go to def and the target is some known extension). You can disable it via `"C_Cpp.autoAddFileAssociations": false`

**Reply by sean-mcmanus** [reply]
@Zylann Let us know if you have a repro for when our extension adds some unwanted file association mapping and so we can check if it's a bug or not. Our extension is somehow thinking an extension of those type is `#included` by another C++ file. I haven't seen a repro or other reports of this.

**Comment by Zylann** [comment]
It happened again. These lines got added to `settings.json`:
```
		"*.idl": "cpp",
		"*.json": "cpp",
		"*.md": "cpp",
		"*.txt": "cpp",
		"*.webp": "cpp"
```
From the modification date, that was 6 hours from now. Unfortunately I can't tell yet what I was doing at that exact moment. Maybe I should constantly record my desktop or write some script that monitors my `settings.json` and plays a horn sound if it ever gets modified

**Comment by Zylann** [comment]
It happened again on my watch this time, I had a script monitoring the file, and this was added:
```json
		"*.idl": "cpp"
```
This was while I was... well, writing code, nothing really special I'm afraid :( 
I was just using Ctrl+V in several places of a C++ file to replace the name of a variable, when it happened out of the blue.
In case it matters, I had 2 other instances of VSCode opened in the background on completely separate projects, all of which I hadn't touched for a while.

I'm afraid I don't know how else I could monitor this with more precision.

**Reply by sean-mcmanus** [reply]
@Zylann It's caused by your use of the Ctrl key while hovering over an identifier for which there exists a matching symbol that is defined in some *.idl file that has been `#included` by a C++ file. VS Code sends us a textDocument/definition message in that case...I'm not sure how we can detect that case versus and actual go to def.

**Reply by Zylann** [reply]
There isn't any idl file in my codebase 🤔or could it be something in the Windows SDK or standard library? 

It's weird because `.idl` files are not something you normally include in a C++ file. The only place I saw one is in the BGFX library, they are a different language used for code generation.
I do have a project using BGFX, but I haven't opened that project in VSCode for a while.

Also I wonder how it came to add all these other extensions from my first post, some of them are really not something that should be `#included` either.

**Reply by sean-mcmanus** [reply]
@Zylann Yeah, it can be from a system header.

**Reply by bobbrow** [reply]
But @sean-mcmanus while this might apply for *.idl files which is an extension we allow, I was pretty sure we wouldn't parse through any other file extensions mentioned in the first comment so an errant goto definition request wouldn't result in us adding those other extensions, right?

@Zylann did you try setting `"C_Cpp.autoAddFileAssociations": false` in your settings yet?  If this is still happening randomly, it would be nice if we could definitively place the blame on our extension or not.

**Reply by Zylann** [reply]
It is still happening randomly. I'll try setting `autoAddFileAssociations` to false.

**Reply by sean-mcmanus** [reply]
@bobbrow No, if a file gets parsed as a source file and it includes an unrecognized file type then it'll add that association. It seems strange for a C/C++ file to include a json file though. We could/should probably add better logging explaining the source for the additional association.
